File: src/main/java/net/sourcebot/api/event/RegisteredListener.java
package net.sourcebot.api.event;

import net.dv8tion.jda.api.events.GenericEvent;
import net.sourcebot.api.module.SourceModule;

import java.util.function.Consumer;

class RegisteredListener<T extends GenericEvent> {
    final SourceModule owner;
    final Consumer<T> consumer;

    RegisteredListener(SourceModule owner, Consumer<T> consumer) {
        this.owner = owner;
        this.consumer = consumer;
    }
}

File: API/src/main/java/net/sourcebot/api/event/RegisteredListener.kt
package net.sourcebot.api.event

import net.sourcebot.api.module.SourceModule

@Suppress("UNCHECKED_CAST")
internal class RegisteredListener<E>(
    internal val module: SourceModule,
    private val listener: (E) -> Unit
) {
    operator fun <T> invoke(event: T) = listener(event as E)
}